### PR TITLE
[RFC] soc: stm32: include HAL headers based on usage

### PR DESCRIPTION
**Pull request by gmarull**
As of today STM32 `soc.h` file includes HAL header files based on driver configuration instead of HAL usage. The proposed changes do the opposite, that is:

- Driver selects which HAL components it needs (`USE_STM32_LL_XXX` or `USE_STM32_HAL_XX`)
- `soc.h` file includes HAL headers based on selected configuration

Advantages:

- It is more clear what a driver needs/uses by looking at `Kconfig`
- If an application uses HAL directly (e.g. because it has a custom driver) it just needs to select `USE_STM32_XXX` and the right headers will be included.

The changes have been only applied to H7 series for demonstration purposes.

### CODE REVIEW

File: drivers/gpio/Kconfig.stm32 [hunk 1]
```diff
@@ -6,6 +6,7 @@
 menuconfig GPIO_STM32
 	bool "GPIO Driver for STM32 family of MCUs"
 	depends on SOC_FAMILY_STM32
+	select USE_STM32_LL_GPIO
```
Comment (jukkar) [inline drivers/gpio/Kconfig.stm32]: Shouldn't we namespace these `USE` options properly? So `STM32_USE_LL_GPIO` instead of `USE_STM32_LL_GPIO`.

Comment (gmarull) [reply drivers/gpio/Kconfig.stm32]: @jukkar that's a good question. I suppose current naming can be changed, even though the purpose of this RFC is to improve how STM32 LL HAL headers are included.

Comment (jukkar) [reply drivers/gpio/Kconfig.stm32]: IMHO the kconfig options should have some kind of namespacing, and in this case the namespace should probably be `STM32` as this option is related to that soc. The `USE_STM32...` looks like an option that might be used by an application in its own kconfig.

Comment (gmarull) [reply drivers/gpio/Kconfig.stm32]: cc: @erwango

Comment (erwango) [reply drivers/gpio/Kconfig.stm32]: I'm open to change this, but I think this is a separate issue.
Also this is not a simple change as I know these symbols have users in  out-of-tree apps. So we would need to notify this change. I propose to focus on current proposal and have a dedicated issue open on this topic.

File: drivers/clock_control/Kconfig.stm32 [hunk 2]
```diff
@@ -8,7 +8,7 @@ menuconfig CLOCK_CONTROL_STM32_CUBE
 	bool "STM32 Reset & Clock Control"
 	depends on SOC_FAMILY_STM32
 	select USE_STM32_LL_UTILS
-	select USE_STM32_LL_RCC if SOC_SERIES_STM32MP1X
+	select USE_STM32_LL_RCC if SOC_SERIES_STM32MP1X || SOC_SERIES_STM32H7X
```
Comment (erwango) [inline drivers/clock_control/Kconfig.stm32]: Miss (for H7)
         select USE_STM32_LL_BUS if SOC_SERIES_STM32H7X
         select USE_STM32_LL_PWR if SOC_SERIES_STM32H7X
         select USE_STM32_LL_SYSTEM if SOC_SERIES_STM32H7X

Comment (gmarull) [reply drivers/clock_control/Kconfig.stm32]: Updated, thanks for noting. Regarding `USE_STM32_LL_PWR` I can't see any reference to `LL_PWR_*` on the H7 clock control driver, so I haven't added it.

**Comment by gmarull** [comment]
@erwango If this approach seems like a way to move forward I can also take care of changing other STM32 series.

**Comment by gmarull** [comment]
@erwango any feedback?

**Comment by andrewboie** [comment]
@ulfalizer do you have any feedback on the Kconfig ontology set up here?

**Review by erwango — CHANGES_REQUESTED** [review]
@gmarull First of all: thanks for the proposal and sorry for the delayed feedback, but I wanted to take some time to think about it.

Today, I'm not in favor of this change. I do get the interest of the change, here are my reasons to push it back today:
- LL API, as provided in STM32 Cube packages, is split in 2 (headers and c). So far it was possible to use only the headers. With your change, it's getting "all or nothing", which is against initial API philosophy.
- It is already possible to use STM32Cube API in your application (https://github.com/zephyrproject-rtos/hal_stm32#use-stm32cube-in-your-application). I'm not personally convinced this facility is not sufficient. I don't think the burden added on the Kconfig files is worth the gain on application side.
- This would add maintenance effort on Zephyr main repo each time STM32Cube API gets updated. Only direct dependency (aside from the potential API breakage) today is `modules/Kconfig.stm32`. This  file should be updated each time we're updating hal_stm32 with new Cube versions. With the proposed change, we would also have to update the `soc.h` files at each update. But this could be automated, so it is a minor point.
- As you may know we're moving drivers to new DT macros. Additionaly, I'm trying to convert stm32 drivers selection from Kconfig to device tree (cf https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/entropy/Kconfig.stm32#L13). Aim is to simplify drivers configuration and minimize dependency with `soc/` section. This change is going in the opposite direction by adding more dependency on Kconfig files from `soc/`. So at the very least, I prefer to complete the current conversion. Then examine again your change and see how it fits.

To sum up: I'm not sure the pain is worth the gain and for sure I don't want to get into this today.

**Comment by gmarull** [comment]
Thanks for your replies @erwango. Here's what I think:

- The API split in headers/sources could be solved by having two sets of config options, e.g. `USE_STM32_LL_TIM` and `USE_STM32_LL_TIM_FULL`. However as a developer I'd expect that all definitions in a header compile and link with no need to enable an extra config option. Note that the non-LL API includes all headers regardless of config as of today (e.g. https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f3xx/drivers/include/stm32f3xx_hal_conf.h#L35). It could be improved with some scripts in HAL module that adjust HAL headers to match Zephyr config.
- It is possible to use the STM32Cube in the application, but it's not easy to create generic STM32 code if I want to use LL API. If my application needs to support two families, e.g. F3 and F4 I'll basically need to have code like this somewhere:

```c
#if defined(CONFIG_SOC_SERIES_STM32F3X)
#include <stm32f3xx_ll_tim.h>
#elif defined(CONFIG_SOC_SERIES_STM32F4X)
#include <stm32f4xx_ll_tim.h>
#endif 
```

- Regarding maintenance, we could also create a `stm32fxxx_ll.h` autogenerated file in the HAL module so that Zephyr is less affected by HAL updates. I could help on that.
- Regarding DT changes I don't see how they are related with this RFC. The driver itself needs to be enabled in Kconfig. DT is useful to describe which units/instances are actually enabled, not if a driver is compiled or not (I understand this from https://docs.zephyrproject.org/latest/guides/dts/dt-vs-kconfig.html#dt-vs-kconfig).

Overall I think that current approach has a dependency in the wrong direction, i.e. SoC includes depend on how drivers are configured. If for example I change a driver or I need to create a new one all SoC headers need to be adjusted, which doesn't feel right...

**Comment by erwango** [comment]
> Thanks for your replies @erwango. Here's what I think:
> 
> * The API split in headers/sources could be solved by having two sets of config options, e.g. `USE_STM32_LL_TIM` and `USE_STM32_LL_TIM_FULL`. However as a developer I'd expect that all definitions in a header compile and link with no need to enable an extra config option. Note that the non-LL API includes all headers regardless of config as of today (e.g. https://github.com/zephyrproject-rtos/hal_stm32/blob/master/stm32cube/stm32f3xx/drivers/include/stm32f3xx_hal_conf.h#L35). It could be improved with some scripts in HAL module that adjust HAL headers to match Zephyr config.

Indeed HAL and LL API differ on that point. This was actually made and kept as is on purpose. I don't have/recall all info why this is standing that way, but that's the way it is in STM32Cube packages and I prefer avoiding messing with these packages and maintaining heavy alternate versions.

> * It is possible to use the STM32Cube in the application, but it's not easy to create generic STM32 code if I want to use LL API. If my application needs to support two families, e.g. F3 and F4 I'll basically need to have code like this somewhere:
> 
> ```c
> #if defined(CONFIG_SOC_SERIES_STM32F3X)
> #include <stm32f3xx_ll_tim.h>
> #elif defined(CONFIG_SOC_SERIES_STM32F4X)
> #include <stm32f4xx_ll_tim.h>
> #endif 
> ```

I get your point. This doesn't feel too much of a burden to me, but I can understand it would be better not having to deal with that.

> 
> * Regarding maintenance, we could also create a `stm32fxxx_ll.h` autogenerated file in the HAL module so that Zephyr is less affected by HAL updates. I could help on that.

Sure. As I said, this point is minor.

> * Regarding DT changes I don't see how they are related with this RFC. The driver itself needs to be enabled in Kconfig. DT is useful to describe which units/instances are actually enabled, not if a driver is compiled or not (I understand this from https://docs.zephyrproject.org/latest/guides/dts/dt-vs-kconfig.html#dt-vs-kconfig).
>

We are actually moving to a more DT controlled model. Kconfig remains in control of the high level SW configuration, but whatever info that could be inferred from dt is removed from Kconfig. Following is an example of where we're moving: https://github.com/zephyrproject-rtos/zephyr/commit/9580992b3e03edda0f3186cd172235338ba19ab2#diff-2dcf3503a60f7c437837c8f32e5b6dc4

From driver Kconfig files perspective, your changes goes in the opposite direction. I don't say it is wrong, but I'd like to complete this step first and then see how your issue can be addressed.
 
> Overall I think that current approach has a dependency in the wrong direction, i.e. SoC includes depend on how drivers are configured. If for example I change a driver or I need to create a new one all SoC headers need to be adjusted, which doesn't feel right...

I get your point and I see the proposal seems reasonable and fixes your immediate need. I just don't want to integrate a change I will have to maintain for years if I'm not comfortable with it. But I do agree to check it once again for next release once the dt changes are well digested and behind us.

**Comment by gmarull** [comment]
@erwango Regarding DT changes, they still would not affect this RFC. The only thing that changes is the default, that is set yo `y` if there's a DT match, but driver software dependencies are still part of Kconfig. If Kconfig/DT is moving to a different direction then I think it would be good to update documentation. cc: @ulfalizer @galak

**Comment by mbolivar-nordic** [comment]
> If Kconfig/DT is moving to a different direction then I think it would be good to update documentation. cc: @ulfalizer @galak @mbolivar-nordic

https://docs.zephyrproject.org/latest/guides/dts/design.html#example-remaining-work

**Comment by gmarull** [comment]
@mbolivar-nordic is stuff like https://github.com/zephyrproject-rtos/zephyr/commit/9580992b3e03edda0f3186cd172235338ba19ab2#diff-2dcf3503a60f7c437837c8f32e5b6dc4 documented? If I understand it correctly, that selects the Kconfig defaults based on DT config, but still does not replace the need for a Kconfig entry, which I think makes sense. There needs to be a way to specify software dependencies.

**Comment by mbolivar-nordic** [comment]
I think it's not super helpful to get hung up on what's documented and what's not in this case.

This is a long story within the project.

Zephyr didn't use to use devicetree at all. Individual peripherals were enabled with kconfig. This scaled badly, and there is general agreement that it didn't work very well, and a new solution was needed.

Along came devicetree, which has the necessary information to record what peripherals are available, their addresses, and whether or not they are enabled.

Now we have a transition problem, because there are existing Kconfig options that were introduced before devicetree was added, which need to be maintained in the short term.

But in the long term, they are going away, which *is* documented. Asking about whether individual driver implementation details for how that transition is going are documented is not super relevant IMO.

**Comment by mbolivar-nordic** [comment]
I think it's not super helpful to get hung up on what's documented and what's not in this case.

This is a long story within the project.

Zephyr didn't use to use devicetree at all. Individual peripherals were enabled with kconfig. This scaled badly, and there is general agreement that it didn't work very well, and a new solution was needed.

Along came devicetree, which has the necessary information to record what peripherals are available, their addresses, and whether or not they are enabled.

Now we have a transition problem, because there are existing Kconfig options that were introduced before devicetree was added, which need to be maintained in the short term.

But in the long term, they are going away, which *is* documented. Asking about whether individual driver implementation details for how that transition is going are documented is not super relevant IMO.

**Comment by gmarull** [comment]
@mbolivar-nordic I think that the question reduces to the following: even though DT is used to select enabled instances (and now to set Kconfig defaults), there needs to be a Kconfig entry for a driver. It is what CMake uses to determine what is built as well as build dependencies (which can't be specified on DT). Is this accurate? I'd like to get this clarified, as I got quite confused after this RFC...

**Comment by gmarull** [comment]
@erwango I've been investigating a bit more, and here are my updated opinions:

- I guess that the reason there is no "universal" header for the LL APIs is because they make extensive use of `static inline` functions. So including them all would increase compile time as all these functions would be *pasted* in each source including the header.
- Either the way it is done now or the proposal on this PR suffer exactly from the same problem. I think that this PR just fix the dependency direction (which I think is good), but does not save us from unnecessary includes.
- One way to solve this problem would be to autogenerate a soc independent header for each LL API and include that header at a driver level. That would produce the minimum compile time and would include what is strictly necessary. Could be done as part of the `hal_stm32` module, so Zephyr is less affected by HAL upgrades. For example:

```c
/** @file stm32_ll_tim.h */
...
#if defined(CONFIG_SOC_SERIES_STM32F3X)
#include <stm32f3xx_ll_tim.h>
#elif defined(CONFIG_SOC_SERIES_STM32F4X)
#include <stm32f4xx_ll_tim.h>
#endif
... 
```

cc: @FRASTM @ABOSTM

**Comment by erwango** [comment]
@gmarull thanks for this feedback and having continued investigations. I'll take few more time to check this. Hope you understand this requires a bit more time than 'usual' PRs.

**Comment by gmarull** [comment]
@erwango Totally understand, changes like this one need to be analyzed in detail, otherwise they could create some unexpected or undesired side effects and so more technical debt...

**Comment by gmarull** [comment]
@erwango any chance this can be discussed again? I ended up creating a similar solution for my out of tree projects. Current Zephyr approach for using/including HAL does not allow to create generic stm32 family code, neither is optimal in the amount of "pasted" inline code.

**Comment by erwango** [comment]
@gmarull feel free to re-open. I haven't forgotten this one.

**Comment by gmarull** [comment]
@erwango let's discuss that after pinctrl stuff is finished
